src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -594,6 +594,21 @@ static uint getDeclarationValEscape(BoundTypeExpression typeExpression, uint val
         public override BoundNode? VisitRecursivePattern(BoundRecursivePattern node)
         {
             SetPatternLocalScopes(node);
+
+            // If the equivalent `Deconstruct` call has unscoped receiver,
+            // we narrow the escape scope of the pattern input
+            // which flows as the escape scope of all ref-struct declaration subpatterns
+            // and so the ref safety of the pattern is equivalent to a `Deconstruct(out var ...)` invocation
+            // where "safe-context inference of declaration expressions" would have the same effect.
+            if (node.DeconstructMethod is { HasUnscopedRefAttribute: true } or
```

Comment:
>{ HasUnscopedRefAttribute: true } [](http://example.com/codeflow?start=42&length=33)

nit: I'm wondering whether this could also be written in terms of `EffectiveScope`, for regularity with second case. Would something like `{ IsExtensionMethod: false, ThisParameter: { EffectiveScope: ScopedKind.None } }` work?

Reply:
`ThisParameter` property looks like it could throw, so added a local function instead of the pattern.